db/mongo/client.go

```diff
@@ -0,0 +1,207 @@
+package mongo
+
+import (
+	"ApniUniversity/config"
+	"ApniUniversity/db"
+	"ApniUniversity/models"
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/spf13/viper"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"time"
+)
+
+const (
+	teacherCollection = "Teacher"
+	subjectCollection = "Subject"
+)
+
+type client struct {
+	connection *mongo.Client
+}
+
+func NewClient(_ db.Options) (db.DataStore, error) {
+	uri := fmt.Sprintf("mongodb://%s:%s", viper.GetString(config.DbHost), viper.GetString(config.DbPort))
+	log().Infof("Initializing MongoDB At: %s", uri)
+	clientOptions := options.Client().ApplyURI(uri)
+	cli, err := mongo.Connect(context.Background(), clientOptions)
+	if err != nil {
+		panic(err)
+	}
+	fmt.Println("Successfully Connected to mongodb!")
+	return &client{connection: cli}, nil
+}
+
+func (c *client) AddTeacher(teacher *models.Teacher) (int, error) {
+	if teacher == nil {
+		return 0, errors.Errorf("Teacher not added!")
+	}
+	teachers, err := c.GetTeachers()
+	if err != nil {
+		return 0, err
+	}
+	if len(teachers) > 0 {
+		teacher.ID = teachers[len(teachers)-1].ID + 1
+	} else {
+		teacher.ID = 1
+	}
+
+	teacher.CreatedAt = time.Now()
+	teacher.UpdatedAt = time.Now()
+
+	collection := c.connection.Database(viper.GetString(config.DbName)).Collection(teacherCollection)
+
+	_, err = collection.InsertOne(context.TODO(), teacher)
+
+	if err != nil {
+		return 0, errors.Wrap(err, "Teacher not added!")
+	}
+	return teacher.ID, nil
+}
+
+func (c *client) GetTeacher(id int) (*models.Teacher, error) {
+	collection := c.connection.Database(viper.GetString(config.DbName)).Collection(teacherCollection)
+	var teacher *models.Teacher
+	err := collection.FindOne(context.TODO(), bson.D{{"_id", id}}).Decode(&teacher)
+	if err != nil {
+		if errors.Is(err, mongo.ErrNoDocuments) {
+			return nil, errors.Wrap(err, "Teacher not Found!")
+		}
+		return nil, err
+	}
+	utc, _ := time.LoadLocation("Asia/Karachi")
+	teacher.CreatedAt = teacher.CreatedAt.In(utc).Round(time.Second)
+	teacher.UpdatedAt = teacher.UpdatedAt.In(utc).Round(time.Second)
+	return teacher, nil
+}
+
+func (c *client) GetTeachers() ([]*models.Teacher, error) {
+	collection := c.connection.Database(viper.GetString(config.DbName)).Collection(teacherCollection)
+	var teachers []*models.Teacher
+	teachersCursor, err := collection.Find(context.TODO(), bson.D{}, options.Find().SetSort(bson.D{{"_id", 1}}))
```

Comment:
why are we hard coding the id?

Reply:
Sir this is sorting based on upon ID. It will return the data based upon ascending order ID